### PR TITLE
Make C headers self contained

### PR DESCRIPTION
**Pull request by Manishearth**
This is a prerequisite for making the C++ headers self contained

You can see this in action at https://github.com/unicode-org/icu4x/pull/2823

This *is* a breaking change for people using the C headers in a way where they relied on the exact import order to get additional types. That is considered bad practice and these are the C headers so I'm fine breaking it.

I'll have to do something similar in C++ though, unsure if it's as kosher there.

### CODE REVIEW

**Comment by Manishearth** [comment]
I'm doing this _before_ HIR migration because _ideally_ I want HIR migration to be a no-op.

File: example/c/include/ICU4XDataProvider_type.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#ifndef ICU4XDataProvider_type_H
```
Comment (sffc) [inline example/c/include/ICU4XDataProvider_type.h:1]: Can you explain why this is superior to putting the forward declarations in the regular header files? I don't think I've seen this pattern very much of having a header file that doesn't have any real declarations in it.

Comment (Manishearth) [reply example/c/include/ICU4XDataProvider_type.h:1]: My original plan for HIR was that the HIR migration should produce no changes to generated code, however I think for the C backend I don't actually care that much.

I do think a nice property of this is that it forces users to import all the headers they need, which is in the long run less brittle.

Comment (sffc) [reply example/c/include/ICU4XDataProvider_type.h:1]: In most cases you're just replacing a pre-existing forward-declaration with the new import of the `_type` file.

Comment (sffc) [reply example/c/include/ICU4XDataProvider_type.h:1]: I guess what I'm saying is, it's fine with me to remove the implicit includes, but it seems like in-file forward declarations are the more common style. Are there examples you've seen of this sort of `_type.h` files in C/C++?

Comment (Manishearth) [reply example/c/include/ICU4XDataProvider_type.h:1]: I haven't; I've seen this pattern in C++ header only situations so I was copying that.

(like I said I think we will need to follow this pattern in C++ unfortunately)

**Comment by Manishearth** [comment]
I think I'm going to close this and directly write a C HIR backend from scratch

**Comment by Manishearth** [comment]
@sffc so, first challenge of just using forward decls: it's not possible to forward-decl just enums

So for example with this pair of headers:

```c
// a.h
#ifndef A_H
#define A_H

#include "b.h"

typedef struct B B;

typedef enum A {
    Foo,
    Bar
} A;

void A_meth(A a, B b);

#endif

// b.h
#ifndef B_H
#define B_H
#include<cstdint>
#include "a.h"

typedef struct B {
    A a;
    int32_t integer;
} B;

void B_meth(A a, B b);

#endif
```

b.h is self contained but a.h is not, since there is no way to forward-declare enums in C. The only way this can be avoided is by forbidding methods on enums. Which makes _some_ sense since in JS at least enums get cast away to strings (but still can be constructed in a typed way!), but I'd rather avoid this step if possible.

**Pull request closed without merging**